gallery_dl/extractor/tiktok.py

```diff
@@ -0,0 +1,207 @@
+# -*- coding: utf-8 -*-
+
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 2 as
+# published by the Free Software Foundation.
+
+"""Extractors for https://www.tiktok.com/"""
+
+from .common import Extractor, Message
+from .. import exception, text, util, ytdl
+from re import compile, escape, IGNORECASE
+from datetime import datetime
+
+BASE_PATTERN = r"(?:https?://)?(?:www\.)?tiktok(?:v?)\.com"
+USER_PATTERN = BASE_PATTERN + r"/+@([\w.]{0,23}\w)?"
+POST_PATTERN = USER_PATTERN + \
+    r"/+(?:[pP][hH][oO][tT][oO]|[vV][iI][dD][eE][oO])/+(?:[0-9]+)/*"
+VM_POST_PATTERN = r"(?:(?:https?://)?(?:(?:vm|vt)\.)?tiktok\.com/+.*/*)|" + \
+    r"(?:(?:https?://)?(?:www\.)?tiktok\.com/+t/+.*/*)"
+SHARE_PATTERN = BASE_PATTERN + r"/+share/+video/+(?:[0-9]+)/*"
+
+
+class TiktokExtractor(Extractor):
+    """Base class for TikTok extractors"""
+
+    category = "tiktok"
+    directory_fmt = ("{category}", "{user}")
+    filename_fmt = "{title} [{id}{index:?_//}{img_id:?_//}].{extension}"
+    archive_fmt = "{id}_{index}_{img_id}"
+    root = "https://www.tiktok.com/"
+    cookies_domain = ".tiktok.com"
+
+    def urls(self):
+        return [self.url]
+
+    def avatar(self):
+        return False
+
+    def items(self):
+        videos = self.config("videos", True)
+        # We assume that all of the URLs served by urls() come from the same
+        # author.
+        downloaded_avatar = not self.avatar()
+        for tiktok_url in self.urls():
+            # If we can recognise that this is a /photo/ link, preemptively
+            # replace it with /video/ to prevent a needless second request.
+            # See below.
+            tiktok_url_to_use = compile(
+                escape("/photo/"),
+                IGNORECASE
+            ).sub("/video/", tiktok_url)
```

Comment:
``` suggestion
            tiktok_url_to_use = \
                compile(r"(?i)/photo/").sub("/video/", tiktok_url)
```

It would honestly be better to just `import re` and use `re.sub()` here
or to pre-compile the regex

or use `tiktok_url.replace()`, although that doesn't handle mixed-case characters, so maybe just do `tiktok_url = tiktok_url.lower()`